WordPress/gutenberg
# Add useSelectors hook for getting a store's selectors

## What?
Create a new function that can be called instead of `useSelect` to get a store's selectors.

```js
export default function useSelectors( storeName ) {
	const registry = useRegistry();

	return registry.select( storeName );
}
```
[Remove complexity](https://github.com/WordPress/gutenberg/pull/40201/files?w=1#diff-842c1e34fc3a1df56dc553d99766a9f86a3e0e141a99632471ed462e2b04035f) from `useSelect` now that the function only does 1 thing.

## Why?
In the previous code, 1 function is used to cover 2 use cases, switching between them with the first parameter's type. This never changes dynamically, code never depends on the function being able to do 2 things.

The original use case is to get data needed during render.

The 2nd use case simply returns a store's selectors, so they can be called in event handlers.

This use case does not need most of the hooks the original use is calling. However because of how hooks work, they needed to be called anyway in this path.

In the new function, the only remaining hook call for getting the controls is to `useRegistry`. All the other logic inside `useSelect` is completely irrelevant.

It also [allows to simplify `useSelect`](https://github.com/WordPress/gutenberg/commit/31b0c2679a63218ed45728021913fd5a60b0ddc5#diff-842c1e34fc3a1df56dc553d99766a9f86a3e0e141a99632471ed462e2b04035f). It won't need any of the checks on the type of argument anymore, and can simply return the map output. For now I kept this out of this branch to make it easier to provide BC.

I didn't test the impact of not having to call these hooks, but I guess it's not 0. Especially if many components use this.

## How?
* Add a new function that only gets the selectors.
* Make all usages of `useSelect` use this function instead.
* Remove complexity from `useSelect` that was only there to make the function do 2 things.

## Testing Instructions
I tried to locate all relevant uses of the function, so depending on test coverage it should be relatively easy to confirm I didn't miss any.

~~Awaiting feedback on what to do with the public API, it could be that `useSelect` would preserve its duplicate function. If that's the case then this change should involve very little risk.~~ For now I indeed kept `useSelect` unchanged so that this PR is unblocked. Perhaps in a follow up it's possible to add the simplified form already and keep a copy of the older one for BC.


## TODO 
The function is exposed with the old signature as a public API. While in practice it should work to check the argument and call the other function, it's being picked up by linting as a rules of hooks violation. Which it technically is, but in practice it only depends on the type which stays stable over time. Still seems like a non optimal solution and a bad example to set.

One option is to leave the old behavior in place in `useSelect`. That would also make it safer in case a usage was missed in this PR.

Alternatively keep an old copy of `useSelect` and export that as a public API. Then internal use could already use the simplified function.
